python/tests/test_diffracted_planewave.py

```diff
@@ -121,17 +138,100 @@ def _pw_amp(x):
       vg_dp = res.vgrp[0]
 
       err = abs(tran_ref-tran_dp)/tran_ref
-      print("tran:, {:2d} (band), {:2d} (order), {:10.8f} (eigensolver), {:10.8f} (planewave), {:10.8f} (error)".format(band,order,tran_ref,tran_dp,err))
+      print("info:, {:2d} (band number), {:2d} (diffraction order), "
+            "{:10.8f} [trans. (band number)], "
+            "{:10.8f} [trans. (diffracted order)], "
+            "{:10.8f} (error)".format(band,
+                                      order,
+                                      tran_ref,
+                                      tran_dp,
+                                      err))
 
       self.assertAlmostEqual(vg_ref,vg_dp,places=5)
       self.assertAlmostEqual(tran_ref,tran_dp,places=5)
 
-  def test_diffracted_planewave(self):
-    self.run_binary_grating_diffraction(2.6,0.4,0.6,0,range(1,6),range(0,5))
-    self.run_binary_grating_diffraction(2.6,0.4,0.6,13.4,range(1,6),[-2,-1,-3,0,-4])
 
-    # self.run_binary_grating_diffraction(10.0,0.5,0.5,0,[2,4,6],[1,3,5])
-    # self.run_binary_grating_diffraction(10.0,0.5,0.5,10.7,[1,4,8],[-6,-4,-2])
+  def run_mode_source(self,m,diffpw):
+    """Computes the transmitted Poynting flux of a
+       binary grating given an incident planewave
+       specified by the diffraction order `m` in the
+       y direction. The incident planewave is defined
+       using a mode source with either a band number
+       or `DiffractedPlanewave` object specified by
+       the boolean flag `diffpw`.
+    """
+    ky = m/self.gp
+    theta = math.asin(ky/(self.fcen*self.ng))
+
+    # k (in source medium) with correct length (plane of incidence: XY)
+    k = mp.Vector3(self.fcen*self.ng).rotate(mp.Vector3(z=1), theta)
+
+    symmetries = []
+    if theta == 0:
+      k = mp.Vector3()
+      self.eig_parity += mp.EVEN_Y
+      symmetries = [mp.Mirror(direction=mp.Y)]
+
+    if diffpw:
+      # the *zeroth* diffraction order defines a planewave with a
+      # wavevector equal to the `k_point` of the `Simulation` object
+      sources = [mp.EigenModeSource(mp.GaussianSource(self.fcen,fwidth=self.df),
+                                    center=self.src_pt,
+                                    size=mp.Vector3(0,self.sy,0),
+                                    eig_band=mp.DiffractedPlanewave((0,0,0),
+                                                                    mp.Vector3(0,1,0),
+                                                                    1,
+                                                                    0))]
+    else:
+      sources = [mp.EigenModeSource(mp.GaussianSource(self.fcen,fwidth=self.df),
+                                    center=self.src_pt,
+                                    size=mp.Vector3(0,self.sy,0),
+                                    direction=mp.NO_DIRECTION,
+                                    eig_kpoint=k,
```

Comment:
This line seems unnecessary — since the source spans the whole width of the cell, it should automatically set the corresponding k component of the eigenmode source, whereas the orthogonal k component will be found by the mode solver.

Reply:
Actually, it turns out that `eig_kpoint=k` *is* necessary because of the line before it: `direction=mp.NO_DIRECTION`. Removing `eig_kpoint` but keeping `direction=mp.NO_DIRECTION` produces this error:
```
meep: invalid direction in add_eigenmode_source
```
This is because setting `direction=mp.NO_DIRECTION` also requires specifying an `eig_kpoint`.

Removing both `eig_kpoint` and `direction=mp.NO_DIRECTION` causes the test to fail because the default `direction=mp.AUTOMATIC` assumes that the `k_point` is always *perpendicular* to the line monitor which it is not. This leads to a wrong mode to be used for the mode decomposition which causes the test to fail:
```
======================================================================
FAIL: test_mode_source (__main__.TestDiffractedPlanewave)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/meep/python/tests/test_diffracted_planewave.py", line 231, in test_mode_source
    self.assertAlmostEqual(tran_bandnum,
AssertionError: 380.9518159531796 != 325.7552498084876 within 4 places (55.19656614469204 difference)
```
The existing setup is therefore correct.

Reply:
> because the default `direction=mp.AUTOMATIC` assumes that the `k_point` is always perpendicular to the line monitor which it is not.

I don't follow — if the monitor region spans the whole width of the cell, for Bloch-periodic boundaries it sets the parallel components of `k` to those of the boundary conditions, and then solves for the perpendicular component, no?

Reply:
I think the issue is related to these lines in `add_source`:

https://github.com/NanoComp/meep/blob/91d36ba73d1ae0eb135d18c783840cff9fcd71ef/python/simulation.py#L2464-L2468

That is, when creating an `EigenModeSource` with the default `direction` of `mp.AUTOMATIC` (which has a value of -1), the `direction` is assigned to the normal direction of the monitor and thus `eig_kpoint` is ignored.

Reply:
The observed behavior is a feature and not a bug. See [#2107 (comment)](https://github.com/NanoComp/meep/pull/2107#discussion_r963118923).